File: windows/src/main/session-store.js
'use strict';

const fs = require('fs');
const path = require('path');
const { EventEmitter } = require('events');
const { app, safeStorage, session: electronSession } = require('electron');
const { PARTITION, MusicClient } = require('./constants');
const { VKAPI } = require('./vk-api');
const { sleep } = require('./util');
const { log } = require('./log');

const FILE = 'session.bin';

function vkSession() {
  return electronSession.fromPartition(PARTITION);
}

async function harvestCookies() {
  const cookies = await vkSession().cookies.get({});
  const useful = cookies.filter((cookie) => {
    const domain = (cookie.domain || '').toLowerCase();
    return domain.includes('vk.com') || domain.includes('vk.ru');
  });
  const hasSession = useful.some((cookie) => cookie.name.includes('remixsid') && cookie.value.length > 10);
  if (!hasSession) return null;
  return useful.map((cookie) => `${cookie.name}=${cookie.value}`).join('; ');
}

async function harvestWithRetry() {
  for (let attempt = 0; attempt < 5; attempt++) {
    const header = await harvestCookies();
    if (header) return header;
    await sleep(250);
  }
  return null;
}

async function clearCookies() {
  const ses = vkSession();
  await ses.clearStorageData();
  await ses.clearCache();
}

// Хранит вход в VK. Токен и куки шифруются средствами ОС (DPAPI на Windows, Keychain на macOS).
class SessionStore extends EventEmitter {
  constructor(api = new VKAPI()) {
    super();
    this.api = api;
    this.session = null;
    this.profile = null;
    this.isBusy = false;
    this.errorMessage = null;
    this.api.onSessionRefreshed = (fresh) => this.adopt(fresh);
    this.restore();
  }

  get isLoggedIn() {
    return this.session !== null;
  }

  snapshot() {
    return {
      loggedIn: this.isLoggedIn,
      profile: this.profile,
      busy: this.isBusy,
      error: this.errorMessage,
      official: Boolean(this.session && this.session.token.startsWith('vk1.'))
    };
  }

  notify() {
    this.emit('change', this.snapshot());
  }

  async finishOAuth(next) {
    this.isBusy = true;
    this.errorMessage = null;
    this.notify();
    try {
      let working = { ...next };
      working.cookieHeader = (await harvestWithRetry()) || working.cookieHeader;
      log('finishOAuth: cookies', working.cookieHeader ? 'found' : 'missing');
      working = await this.api.resolveMusicSession(working);
      log('finishOAuth: music token received');
      let user;
      try {
        user = await this.api.profile(working);
      } catch {
        user = { id: working.userId, firstName: 'VK', lastName: '', photoURL: null };
      }
      this.session = working;
      this.profile = user;
      this.persist();
      log('finishOAuth: session saved');
    } catch (error) {
      log('finishOAuth failed', error.message);
      this.errorMessage = error.message;
    } finally {
      this.isBusy = false;
      this.notify();
    }
  }

  async prepareAudioSession() {
    if (!this.session) return;
    const current = { ...this.session };
    if (!current.cookieHeader) current.cookieHeader = await harvestWithRetry();
    try {
      const upgraded = await this.api.resolveMusicSession(current);
      const changed = JSON.stringify(upgraded) !== JSON.stringify(this.session);
      this.session = upgraded;
      if (changed) {
        try {
          this.profile = await this.api.profile(upgraded);
        } catch {
          // Оставляем прежний профиль.
        }
        this.persist();
      }
      if (upgraded.token.startsWith('vk1.')) this.errorMessage = null;
    } catch (error) {
      log('prepareAudioSession failed', error.message);
      this.errorMessage = error.message;
    }
    this.notify();
  }

  // Обновлённый токен музыки подхватываем сразу, чтобы не запрашивать его заново при каждом вызове API.
  adopt(fresh) {
    if (!this.session || fresh.token === this.session.token) return;
    this.session = fresh;
    this.persist();
  }

  clearError() {
    this.errorMessage = null;
  }

  async logout() {
    this.session = null;
    this.profile = null;
    this.errorMessage = null;
    this.api.clearCaches();
    this.deleteStored();
    this.notify();
    await clearCookies();
  }

  // ---- Хранилище ----

  filePath() {
    return path.join(app.getPath('userData'), FILE);
  }

  persist() {
    if (!this.session || !this.profile) return;
    if (!safeStorage.isEncryptionAvailable()) return;
    try {
      const data = JSON.stringify({ session: this.session, profile: this.profile });
      fs.writeFileSync(this.filePath(), safeStorage.encryptString(data));
    } catch {
      // Без сохранения вход просто не переживёт перезапуск.
    }
  }

  restore() {
    try {
      if (!safeStorage.isEncryptionAvailable()) return;
      const raw = fs.readFileSync(this.filePath());
      const stored = JSON.parse(safeStorage.decryptString(raw));
      if (stored && stored.session && stored.profile) {
        if (!stored.session.userAgent) stored.session.userAgent = MusicClient.userAgent;
        this.session = stored.session;
        this.profile = stored.profile;
      }
    } catch {
      // Нет сохранённого входа.
    }
  }

  deleteStored() {
    try {
      fs.rmSync(this.filePath(), { force: true });
    } catch {
      // Файла может не быть.
    }
  }
}

module.exports = { SessionStore, harvestCookies, clearCookies, vkSession };

File: Fenura/SessionStore.swift
import Foundation
import Combine
import WebKit

enum VKCookieJar {
    nonisolated static func harvest() async -> String? {
        await withCheckedContinuation { continuation in
            let finish: ([HTTPCookie]) -> Void = { cookies in
                let useful = cookies.filter { cookie in
                    let domain = cookie.domain.lowercased()
                    return domain.contains("vk.com") || domain.contains("vk.ru")
                }
                guard useful.contains(where: { $0.name.contains("remixsid") && $0.value.count > 10 }) else {
                    continuation.resume(returning: nil)
                    return
                }
                let header = useful.map { "\($0.name)=\($0.value)" }.joined(separator: "; ")
                continuation.resume(returning: header)
            }

            DispatchQueue.main.async {
                WKWebsiteDataStore.default().httpCookieStore.getAllCookies(finish)
            }
        }
    }

    nonisolated static func harvestWithRetry() async -> String? {
        for _ in 0..<5 {
            if let header = await harvest() { return header }
            try? await Task.sleep(nanoseconds: 250_000_000)
        }
        return nil
    }

    nonisolated static func clear() {
        DispatchQueue.main.async {
            let store = WKWebsiteDataStore.default()
            store.fetchDataRecords(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes()) { records in
                let vk = records.filter { record in
                    let name = record.displayName.lowercased()
                    return name.contains("vk.") || name.contains("vk.com") || name.contains("vk.ru")
                }
                store.removeData(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes(), for: vk, completionHandler: {})
            }
        }
    }
}

struct PersistedSession: Codable {
    let session: VKSession
    let profile: UserProfile
}

@MainActor
final class SessionStore: ObservableObject {
    private let api = VKAPI()

    @Published var session: VKSession?
    @Published var profile: UserProfile?
    @Published var challenge: AuthChallenge?
    @Published var isBusy = false
    @Published var errorMessage: String?

    var isLoggedIn: Bool { session != nil }

    init() {
        restore()
    }

    func finishOAuth(_ next: VKSession) async {
        isBusy = true
        errorMessage = nil
        defer { isBusy = false }

        do {
            var working = next
            working.cookieHeader = await VKCookieJar.harvestWithRetry() ?? working.cookieHeader
            working = try await api.resolveMusicSession(working)
            let user = (try? await api.profile(session: working)) ?? UserProfile(
                id: working.userId,
                firstName: "VK",
                lastName: "",
                photoURL: nil
            )
            session = working
            profile = user
            challenge = nil
            persist()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func prepareAudioSession() async {
        guard var current = session else { return }
        if current.cookieHeader == nil {
            current.cookieHeader = await VKCookieJar.harvestWithRetry()
        }
        do {
            let upgraded = try await api.resolveMusicSession(current)
            if upgraded != current {
                session = upgraded
                if let user = try? await api.profile(session: upgraded) {
                    profile = user
                }
                persist()
            } else {
                session = upgraded
                api.remember(upgraded)
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func clearError() {
        errorMessage = nil
    }

    func logout() {
        session = nil
        profile = nil
        challenge = nil
        errorMessage = nil
        KeychainStore.delete()
        VKCookieJar.clear()
    }

    func apiClient() -> VKAPI { api }

    private func persist() {
        guard let session, let profile else { return }
        if let data = try? JSONEncoder().encode(PersistedSession(session: session, profile: profile)) {
            KeychainStore.save(data)
        }
    }

    private func restore() {
        guard let data = KeychainStore.load(),
              let stored = try? JSONDecoder().decode(PersistedSession.self, from: data) else { return }
        session = stored.session
        profile = stored.profile
        api.remember(stored.session)
    }
}
